src/context.jl

```diff
@@ -1,24 +1,77 @@
 # OpenCL.Context
 
+const _ctx_reference_count = Dict{CL_context, Int}()
+
+
+function create_jl_reference!(ctx_id::CL_context)
+    if haskey(_ctx_reference_count, ctx_id) # for the first jl reference, we already have a refcount of 1
+        @check api.clRetainContext(ctx_id) # increase internal refcount, if creating an additional reference
+    end
+    refcount = get!(_ctx_reference_count, ctx_id, 0)
+    _ctx_reference_count[ctx_id] = refcount + 1
+    return
+end
+function free_jl_reference!(ctx_id::CL_context)
+    if !haskey(_ctx_reference_count, ctx_id)
+        error("Freeing unknown context")
+    end
+    refcount = _ctx_reference_count[ctx_id]
+    if refcount == 0
+        error("Double free of context id: ", ctx_id)
+    elseif refcount == 1
+        delete!(_ctx_reference_count, ctx_id)
+        return
+    end
+    _ctx_reference_count[ctx_id] = refcount - 1
+    return
+end
+
 type Context <: CLObject
```

Comment:
The other option would be to turn Context into singelton objects... similar to how we handle `CuContext` in the CUDA packages since reference counting for Julia GC managed objects is possible but might be inaccurate if the finalizers are delayed.